### PR TITLE
Add continuum damage to the crystal plasticity system

### PR DESCRIPTION
**Pull request by reverendbedford**
This PR adds a system for including continuum damage in crystal plasticity models.  It implements an alternative kinematic formulation that degrades the elasticity tensor used in the CP stress update using a projection operator, which depends on the stress and some set of internal variables.

The specific models implemented in the general framework damage the elasticity tensor on a plane-by-plane basis, where the stiffness in the normal and shear directions of each slip plane can be degraded separately.

### CODE REVIEW

File: src/cp/crystaldamage.cxx [hunk 1]
```diff
@@ -0,0 +1,631 @@
...
+  return pset;
+}
+
+double SigmoidTransformation::map(double damage, double normal_stress)
+{
+  if (damage < 0)
+    return 0;
+  else if (damage < c_)
+  {
+    double dtrial = 1.0/(1.0 + std::pow(damage / (c_ - damage), -beta_));
```
Comment (ajey091) [inline src/cp/crystaldamage.cxx:554]: This is fine, but can be written more simply as 
`double dtrial = std::pow((c_ - damage) / damage ,beta_)`

File: src/cp/crystaldamage.cxx [hunk 2]
```diff
@@ -0,0 +1,631 @@
...
+  }
+  return cut_;
+}
+
+double SigmoidTransformation::d_map_d_damage(double damage, double normal_stress)
+{
+  if (damage < 0)
+    return 0;
+  else if (damage < c_) {
+    double dtrial = 1.0/(1.0 + std::pow(damage / (c_ - damage), -beta_));
```
Comment (ajey091) [inline src/cp/crystaldamage.cxx:568]: Same as above.

File: doc/sphinx/cp/cpdmg/PlanarDamageModel.rst [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+.. math::
+
+   S_{ijkl}^{\left(i\right)}=\left(\delta_{ik}-n_{i}^{\left(i\right)}n_{k}^{\left(i\right)}\right)n_{j}^{\left(i\right)}n_{l}^{\left(i\right)}.
+
+The stress normal to the plane can be calculated as
+
+.. math::
+
+   \sigma_{\bot}^{\left(i\right)}=\sigma_{ij}n_{i}^{\left(i\right)}n_{j}^{\left(i\right)}.
+
```
Comment (ajey091) [inline doc/sphinx/cp/cpdmg/PlanarDamageModel.rst:43]: Maybe it's useful to include the description of the damage metric used and also provide the source for the same.

File: test/regression/reference.xml [hunk 4]
```diff
@@ -1,4 +1,55 @@
...
+      </dmodel>
+    </kinematics>
+    <lattice type="CubicLattice">
+      <a>1.0</a>
+      <slip_systems>
+        1 1 0 ; 1 1 1
+      </slip_systems>
+    </lattice>
+  </cpdmg>
+
```
Comment (ajey091) [inline test/regression/reference.xml:52]: This section should be updated to a newer version of the material file with corrections:
`<materials>
  <cpdmg type="SingleCrystalModel">
    <miter>15</miter>
    <max_divide>10</max_divide>
    <verbose>false</verbose>
    <kinematics type="DamagedStandardKinematicModel">
      <emodel type="IsotropicLinearElasticModel">
        <m1_type>youngs</m1_type>
        <m1>150000</m1>
        <m2_type>poissons</m2_type>
        <m2>0.285</m2>
      </emodel>
      <imodel type="AsaroInelasticity">
        <rule type="PowerLawSlipRule">
          <resistance type="VoceSlipHardening">
            <tau_sat>12</tau_sat>
            <b>66.6666666667</b>
            <tau_0>40</tau_0>
          </resistance>
          <gamma0>9.55470706737e-08</gamma0>
          <n>12</n>
        </rule>
      </imodel>
      <dmodel type="PlanarDamageModel">
        <damage type="WorkPlaneDamage"/>
        <shear_transform type="SigmoidTransformation">
          <c>100.0</c>
          <beta>2.0</beta>
          <cut>0.25</cut>
        </shear_transform>
        <normal_transform type="SigmoidTransformation">
          <c>100.0</c>
          <beta>2.0</beta>
          <cut>0.25</cut>
        </normal_transform>
        <lattice type="CubicLattice">
          <a>1.0</a>
          <slip_systems>
            1 1 1 ; 1 1 0
          </slip_systems>
        </lattice>
      </dmodel>
    </kinematics>
    <lattice type="CubicLattice">
      <a>1.0</a>
      <slip_systems>
        1 1 1 ; 1 1 0
      </slip_systems>
    </lattice>
  </cpdmg>
</materials>`

**Review by ajey091 — CHANGES_REQUESTED** [review]
The only major change would be to update the material description in the xml file. I have left a few comments for minor changes.

File: src/cp/crystaldamage.cxx (hunk 1 above)
Comment (reverendbedford) [reply src/cp/crystaldamage.cxx:554]: I can get to `1.0 / (1.0 + std::pow(c_ / damage -1, beta_);` but not all the way to what you have.  I made at least that simplification.

File: src/cp/crystaldamage.cxx (hunk 2 above)
Comment (reverendbedford) [reply src/cp/crystaldamage.cxx:568]: Same

File: doc/sphinx/cp/cpdmg/PlanarDamageModel.rst (hunk 3 above)
Comment (reverendbedford) [reply doc/sphinx/cp/cpdmg/PlanarDamageModel.rst:43]: The damage metrics are listed below this line in the `SlipPlameDamage` TOC.

File: test/regression/reference.xml (hunk 4 above)
Comment (reverendbedford) [reply test/regression/reference.xml:52]: This was obviated by changes on another branch.

**Review by ajey091 — APPROVED** [review]

**Pull request merged**